polkadot/xcm/xcm-builder/src/barriers.rs

```diff
@@ -502,3 +515,125 @@ impl DenyExecution for DenyReserveTransferToRelayChain {
+	Outer: DenyExecution,
+	Inner: DenyExecution;
+
+impl<Outer, Inner> DenyNestedXcmInstructions<Outer, Inner>
+where
+	Outer: DenyExecution,
+	Inner: DenyExecution,
+{
+	// Recursively applies the deny filter to a nested XCM.
+	///
+	/// This function ensures that restricted instructions are blocked at any depth within the XCM.
+	/// It maintains a **recursion counter** to prevent stack overflows due to a deep nesting.
+	fn deny_recursively<RuntimeCall>(
+		origin: &Location,
+		nested_xcm: &mut Xcm<RuntimeCall>,
+		max_weight: Weight,
+		properties: &mut Properties,
+	) -> Result<Result<ControlFlow<()>, ProcessMessageError>, ProcessMessageError> {
+		// Apply the `Inner` deny filter to the nested XCM.
+		let _ = Inner::deny_execution(origin, nested_xcm.inner_mut(), max_weight, properties)
+			.inspect_err(|e| {
+				log::warn!(
+					target: "xcm::barriers",
+					"Execution denied by Inner filter, origin: {:?}, nested_xcm: {:?}, error: {:?}",
+					origin, nested_xcm, e
+				);
+			})?;
+
+		// Initialise the recursion count the first time we enter recursive execution.
+		let _ = recursion_count::using_once(&mut 1, || {
+			recursion_count::with(|count| {
+				if *count > xcm_executor::RECURSION_LIMIT {
+					log::error!(
+						target: "xcm::barriers",
+						"Recursion limit exceeded, origin: {:?}, nested_xcm: {:?}, count: {count}",
+						origin, nested_xcm
+					);
+
+					return Err(ProcessMessageError::StackLimitReached);
+				}
+				*count = count.saturating_add(1);
+				Ok(())
+			})
+			// Fallback safety in case of an unexpected failure.
+			.unwrap_or(Ok(()))?;
+
+			// Ensure that we always decrement the counter after processing.
+			sp_core::defer! {
+				recursion_count::with(|count| {
+					*count = count.saturating_sub(1);
+				});
+			}
+
+			// Recursively check the nested XCM instructions.
+			Outer::deny_execution(origin, nested_xcm.inner_mut(), max_weight, properties)
+		})?;
+
+		Ok(Ok(ControlFlow::Continue(())))
+	}
+}
+
+/// Denies execution if the XCM contains restricted **local** instructions, first checking at the
+/// top-level and then **recursively** using `DenyNestedXcmInstructions`.
+///
+/// This barrier only applies to **locally executed** XCM instructions (`SetAppendix`,
+/// `SetErrorHandler`, and `ExecuteWithOrigin`). Remote parts of the XCM are expected to be
+/// validated by receiving chain's barrier.
+///
+/// Note: Ensures that restricted instructions do not execute on the local chain, enforcing stricter
+/// execution policies, while allowing remote chains to enforce their own rules.
+pub struct DenyNestedLocalInstructions<Inner>(PhantomData<Inner>);
+
+impl<Inner: DenyExecution> DenyExecution for DenyNestedLocalInstructions<Inner> {
+	/// Denies execution of restricted local nested XCM instructions.
+	///
+	/// This checks for `SetAppendix`, `SetErrorHandler`, and `ExecuteWithOrigin` instruction
+	/// applying the deny filter **recursively** to any nested XCMs found.
+	fn deny_execution<RuntimeCall>(
```

Comment:
I would rename `DenyNestedLocalInstructions` -> `DenyLocalInstructions`, because it does top-level + nested

```suggestion
pub struct DenyLocalInstructions<Inner>(PhantomData<Inner>);

impl<Inner: DenyExecution> DenyExecution for DenyLocalInstructions<Inner> {
	/// Denies execution of restricted local XCM instructions, including nested XCM within the `SetAppendix`, `SetErrorHandler`, and `ExecuteWithOrigin` instructions. Applies the `Inner` deny filter **recursively** to any nested XCMs found.
	fn deny_execution<RuntimeCall>(
```

Reply:
Renamed to `DenyLocalInstructions`